chain/network/src/peer/peer_actor.rs

```diff
@@ -979,6 +691,48 @@ impl PeerActor {
         }
     }
 
+    fn receive_message(
+        &self,
+        ctx: &mut actix::Context<Self>,
+        conn: &connection::Connection,
+        msg: PeerMessage,
+    ) {
+        // This is a fancy way to clone the message iff event_sink is non-null.
+        // If you have a better idea on how to achieve that, feel free to improve this.
+        let message_processed_event = self
+            .network_state
+            .config
+            .event_sink
+            .delayed_push(|| Event::MessageProcessed(msg.clone()));
+        let was_requested = match &msg {
+            PeerMessage::Block(block) => {
+                let hash = *block.hash();
+                conn.chain_height.fetch_max(block.header().height(), Ordering::Relaxed);
+                self.tracker.lock().push_received(hash);
+                self.tracker.lock().has_request(&hash)
+            }
+            _ => false,
+        };
+        let clock = self.clock.clone();
+        let network_state = self.network_state.clone();
+        let peer_id = conn.peer_info.id.clone();
+        ctx.spawn(
+            wrap_future(async move {
+                network_state.receive_message(&clock, peer_id, msg, was_requested).await
+            })
+            .then(|res, act: &mut PeerActor, ctx| {
+                match res {
+                    // TODO(gprusak): make sure that for routed messages we drop routeback info correctly.
+                    Ok(Some(resp)) => act.send_message_or_log(&resp),
+                    Ok(None) => {}
+                    Err(ban_reason) => act.stop(ctx, ClosingReason::Ban(ban_reason)),
+                }
+                message_processed_event();
+                wrap_future(async {})
+            }),
```

Comment:
```suggestion
            .map(|res, act: &mut PeerActor, ctx| {
                match res {
                    // TODO(gprusak): make sure that for routed messages we drop routeback info correctly.
                    Ok(Some(resp)) => act.send_message_or_log(&resp),
                    Ok(None) => {}
                    Err(ban_reason) => act.stop(ctx, ClosingReason::Ban(ban_reason)),
                }
                message_processed_event();
            }),
```

Reply:
are you sure? This will be a map() with side effects then.

Reply:
Good question. I think I am: this is a future, not an option. So, there's no *real* side effect here: what `.map` does is just creating a `Map` future, the side-effect gets executed only when the thing is `.awaited`. 

This is in contrast to options, where `opt.map(|val| side_effect(val))` *looks* like it postpones the computation, but actually immediately runs the thing. 

Of course the first-best solution here would be to avoid high-order functions and just use `.await`, but we can't do that until we move away from actor futures.